### PR TITLE
Remove synchronized in PartitioningExchanger

### PR DESCRIPTION
**Pull request by fmeiser**
This PR removes `synchronized ` in `PartitioningExchanger`. I guess `synchronized` was used to synchronize to reuse allocated partition assignments. This reusing is replaced by createing partition assignments in expected size.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/exchange/PartitioningExchanger.java [hunk 1]
```diff
@@ -61,19 +59,17 @@ public PartitioningExchanger(
...
-    public synchronized void accept(Page page)
+    public void accept(Page page)
     {
-        // reset the assignment lists
-        for (IntList partitionAssignment : partitionAssignments) {
-            partitionAssignment.clear();
+        int partitionCount = buffers.size();
+
+        IntArrayList[] partitionAssignments = new IntArrayList[partitionCount];
+        int expectedSize = page.getPositionCount() / partitionCount;
```
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/operator/exchange/PartitioningExchanger.java]: Unrelated change (separate commit maybe?).

Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/operator/exchange/PartitioningExchanger.java]: Is it correct to assume even distribution here?
It at least specifies some capacity instead of relying on `IntArrayList` defaults.

**Comment by maciejgrzybek** [comment]
Looks reasonable given no mutable context is shared among concurrent calls to `accept`.
Though, it would be worth to ask @dain for opinion here.

**Comment by dain** [comment]
Did you find a performance improvement from this change?

There are two separate things you are changing here:
1) synchronization - This actually isn't necessary as the `PartitionExchanger` is guaranteed to be used in thread safe context (it is only used by a single `Operator` instance)
2) reuse of the arrays - It is possible that creating new arrays each time is more efficient (or doesn't matter).  My guess is the array creation has no impact so we may want to simply recreate every time to simplify the code. This is why I'd curious if you saw a performance boost.

**Comment by fmeiser** [comment]
@dain There was no influence on performance. My motivation was to simplify the code and to allocate resources as short as possible.

**Comment by dain** [comment]
For now, I'm going to suggest we leave this alone, mostly because #5877 makes many changes to the local exchange code.  In the future we should write a benchmark for this to see the impact of the allocations.  Additionally, we might be better off using a PageBuilder for each partition to avoid creating small pages, but that comes at the cost of more buffer memory.  Alternatively, we could introduce a MaskedBlock to avoid the copy, but that might have downstream performance effects.

**Comment by fmeiser** [comment]
@dain Do we already have a MaskedBlock? I guess if we use MaskedBlock, we have to introduce a reference counting for blocks to calculate the correct memory usage.

**Comment by dain** [comment]
We do not have a MaskedBlock yet.  We have considered it in the past, but we have found extra layers of indirection like this can really slow everything down.  We do this layering technique in RLE and Dictionary blocks, but those are handled specially.  I'd be more comfortable approaching this change when we have a continuous performance analysis system in place.

**Comment by fmeiser** [comment]
Okay. MaskedBlock will also increase memory cache misses in downstream processing.
